pkg/credentialprovider/plugin/plugin.go

```diff
@@ -0,0 +1,393 @@
+	}
+
+	if found {
+		return cachedConfig
+	}
+
+	// fetch credentials from the exec plugin if the image is allowed and not already matched against cached objects
+	response, err := p.plugin.ExecPlugin(image)
+	if err != nil {
+		klog.Errorf("Failed getting credential from external registry credential provider: %v", err)
+		return credentialprovider.DockerConfig{}
+	}
+
+	if len(response.Auth) == 0 {
+		return credentialprovider.DockerConfig{}
+	}
+
+	var cacheKey string
+	switch cacheKeyType := response.CacheKeyType; cacheKeyType {
+	case criauth.ImagePluginCacheKeyType:
+		cacheKey = image
+	case criauth.RegistryPluginCacheKeyType:
+		registry, err := parseRegistry(image)
+		if err != nil {
+			klog.Errorf("error parsing registry from image %s: %v", image, err)
+			return credentialprovider.DockerConfig{}
+		}
+
+		cacheKey = registry
+	case criauth.GlobalPluginCacheKeyType:
+		cacheKey = globalCacheKey
+	}
+
+	dockerConfig := make(credentialprovider.DockerConfig, 0)
+	for matchImage, authConfig := range response.Auth {
+		dockerConfig[matchImage] = credentialprovider.DockerConfigEntry{
+			Username: authConfig.Username,
+			Password: authConfig.Password,
+			Email:    "plugin@criauth.k8s.io", // email is deprecated, this is just a dummy email
+		}
+	}
+
+	// Don't cache response if cache key type was not specified, return dockerConfig.
+	if cacheKey == "" {
+		return dockerConfig
+	}
+
+	// cache duration was explicitly 0 so don't cache this response at all.
+	if response.CacheDuration != nil && response.CacheDuration.Duration == 0 {
+		return dockerConfig
+	}
+
+	var expiresAt time.Time
+	// nil cache duration means use the default cache duration
+	if response.CacheDuration == nil {
+		expiresAt = time.Now().Add(p.defaultCacheDuration)
+	} else {
+		expiresAt = time.Now().Add(response.CacheDuration.Duration)
+	}
+
+	cachedEntry := &cacheEntry{
+		key:         cacheKey,
+		credentials: dockerConfig,
+		expiresAt:   expiresAt,
+	}
+
+	if err := p.cache.Add(cachedEntry); err != nil {
+		klog.Errorf("Error adding auth entry to cache: %v", err)
+	}
+
+	return dockerConfig
+}
+
+// Enabled always returns true since registration of the plugin via kubelet implies it should be enabled.
+func (e *pluginProvider) Enabled() bool {
+	return true
+}
+
+// IsPlugin returns true since all external credential providers are plugins
+func (e *pluginProvider) IsPlugin() bool {
+	return true
+}
+
+// isImageAllowed returns true if the image matches against the list of allowed matches by the plugin.
+func (p *pluginProvider) isImageAllowed(image string) bool {
+	for _, matchImage := range p.matchImages {
+		if matched, _ := credentialprovider.URLsMatchStr(matchImage, image); matched {
+			return true
+		}
+	}
+
+	return false
+}
+
+// getCachedCredentials returns a credentialprovider.DockerConfig if cached from the plugin.
+func (p *pluginProvider) getCachedCredentials(image string) (credentialprovider.DockerConfig, bool, error) {
```

Comment:
go from most-specific to least-specific:

1. image
2. registry
3. global

Reply:
done